src/utilities/coerceInputValue.js

```diff
@@ -48,15 +69,35 @@ function defaultOnError(
   throw error;
 }
 
+function inputCoercion(type: GraphQLLeafType, value: mixed): mixed {
+  // Scalars and Enums determine if a external input value is valid via
+  // parseValue(),
+  return type.parseValue(value);
+}
+
+function defaultValueCoercion(type: GraphQLLeafType, value: mixed): mixed {
+  // Default values are initially represented as internal values, "serialize"
+  // converts the internal value to an external value, and "parseValue" converts
+  // back to an internal value.
+  return type.parseValue(type.serialize(value));
```

Comment:
A key example of this is Enums, where the external (string) and internal (any) representations differ. Custom scalars can have the same property

Reply:
I don't this this is correct:
`serialize` is for coercing the values of resolved values (the return values from Resolver). It takes the result of a resolver and produces an result value ("result coercion" is the more correct term here).
`parseValue` is for coercing input runtime values (programming language specific values). The two use cases for that kind of coercing are variables values and default values. A `GraphQLInputField` contains a `defaultValue` which needs to be coerced via `parseValue`.

Reply:
That would be a change in behaviour for graphql.js, but I think that change is desired - specifying that the default for an enum is `"MY_CUSTOM_SORTER"` rather than `require('sorters').myCustomSorter` (which may happen to be a function) makes sense to me - it's more in line with specifying the value via JSON within a variable.

I think Lee's solution may work to maintain the current behaviour - if you return an enum value (which may happen to be a function) from a resolver then that gets converted via `serialize`; similarly if you have the exact same enum value (i.e. a function) within `defaultValue` then using serialize to turn it back into the GraphQL enum makes sense to me.

Reply:
Let me expand a bit, why I think this can't work in general.

The possible output values of `serialize` can be completely disjunct from the input values of `parseValue`. They simply have an orthogonal purpose. `parseValue` will represent an "internal input value" and should accept a wide range of different inputs and return a good representation of the value to be received as input for DataFetcher.
The output of `serialize` on the other hand is the final result of a field. 
In practice most Scalars probably accepts `serialize` values as input for `parseValue`, but this is not required at all.
I could define a `Currency` scalar for example  which has a special `CurrencyOutput` as the result of `serialize` which is not accepted as input for `parseValue`.

Also there is nothing that actually forces you to implement to `serialize` in the first place. See https://github.com/jaydenseric/graphql-upload/blob/master/public/GraphQLUpload.js for an example.

Reply:
Thanks for clarifying Andi; that makes sense.